url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_BigInteger.java

```diff
@@ -0,0 +1,617 @@
+     * @return 1/x Accurate to decimalPlace number of decimal places. If x = 0
+     * then an IllegalArgumentException is thrown
+     */
+    public static BigDecimal reciprocal(
+            BigInteger x,
+            int decimalPlaces,
+            RoundingMode a_RoundingMode) {
+        if (x.compareTo(BigInteger.ZERO) == 0) {
+            throw new IllegalArgumentException(
+                    "x = 0 in " + Generic_BigInteger.class
+                    + ".reciprocal(BigInteger,int,RoundingMode)");
+        }
+        BigDecimal result = BigDecimal.ONE.divide(
+                new BigDecimal(x),
+                decimalPlaces,
+                a_RoundingMode);
+        return Generic_BigDecimal.roundIfNecessary(result, decimalPlaces, a_RoundingMode);
+    }
+
+    /**
+     * e^y = 1 + y/1! + y^2/2! + y^3/3! +...
+     * @param y
+     * @param a_Generic_BigDecimal
+     * @param decimalPlaces
+     * @param a_RoundingMode
+     * @return e^y where e is the Euler constant to a sufficient precision to
+     * return the result accurate to the requested precision.
+     */
+    protected static BigDecimal exp(
+            BigInteger y,
+            Generic_BigDecimal a_Generic_BigDecimal,
+            int decimalPlaces,
+            RoundingMode a_RoundingMode) {
+        // Deal with special cases
+        if (y.compareTo(BigInteger.ZERO) == 0) {
+            return BigDecimal.ONE;
+        }
+        BigDecimal y_BigDecimal = new BigDecimal(y);
+        return Generic_BigDecimal.exp(
+                y_BigDecimal,
+                a_Generic_BigDecimal,
+                decimalPlaces,
+                a_RoundingMode);
+    }
+
+    /**
+     * There are methods to get large random numbers.
+     * This method would only give un-bias results for upperLimit being a number 
+     * made of only 9's such as 99999. The problem with any other number is that 
+     * there is an uneven distribution of decimal digits skewed towards 1.
+     * There are many more 1's in the numbers 0 to 1234567 than there are any 
+     * other number, second most common is 2 and so on. For any range of numbers
+     * this distribution is different.
+     * There is a constructor method for BigDecimal that supports this, but only
+     * for uniform distributions over a binary power range.
+     * @param upperLimit
+     * @return a random integer as a BigInteger between 0 and upperLimit
+     * inclusive
+     */
+    public BigInteger getRandom(
+            BigInteger upperLimit) {
+        return getRandom(get_RandomArrayMinLength(1)[0],upperLimit);
+    }
+
+    public BigInteger getRandom(
+            Random a_Random, 
+            BigInteger upperLimit) {
+        // Special cases
+        if (upperLimit.compareTo(BigInteger.ZERO) == 0) {
+            return BigInteger.ZERO;
+        }
+        if (upperLimit.compareTo(BigInteger.valueOf(Integer.MAX_VALUE - 1)) == -1) {
+        int randomInt = a_Random.nextInt(upperLimit.intValue() + 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;a_Random&#39; (index: 0 | type: Random). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;upperLimit&#39;. The same parameter type is validated here for example: url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/utilities/Generic_Collections.java:198. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).